File: solutions/0054_Spiral_Matrix/iter_by_layer.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ret = new ArrayList<>();
        if (matrix.length == 0) {
            return ret;
        }
        int l = 0;
        int layer = (Math.min(matrix.length, matrix[0].length) + 1) / 2;
        while (l < layer) {
            int r1 = l;
            int r2 = matrix.length - 1 - l;
            int c1 = l;
            int c2 = matrix[0].length - 1 - l;
            // last layer is a horizontal line
            for (int c = c1; c <= c2; c++) ret.add(matrix[r1][c]);
            // last layer is a verticle line
            for (int r = r1 + 1; r <= r2; r++) ret.add(matrix[r][c2]);
            if (r1 < r2 && c1 < c2) {
                for (int c = c2 - 1; c > c1; c--) ret.add(matrix[r2][c]);
                for (int r = r2; r > r1; r--) ret.add(matrix[r][c1]);
            }
            l += 1;
        }
        return ret;
    }
}

File: solutions/0059_Spiral_Matrix_II/iter_by_layer.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        ret = [[0 for _ in range(n)] for _ in range(n)]
        v = 1
        l = 0
        while l < (n+1) // 2:
            r1, r2, c1, c2 = l, n-1-l, l, n-1-l
            for i in range(c1, c2+1, 1):
                ret[r1][i] = v
                v += 1
            for i in range(r1+1, r2+1, 1):
                ret[i][c2] = v
                v += 1
            if c2 > c1:
                for i in range(c2-1, c1-1, -1):
                    ret[r2][i] = v
                    v += 1
            if r2 > r1:
                for i in range(r2-1, r1, -1):
                    ret[i][c1] = v
                    v += 1
            l += 1
        return ret
